### PR TITLE
upgrade to graphql v17 including support for new incremental delivery format

### PR DESCRIPTION
**Pull request by yaacovCR**
This PR provides support for the next iteration of the incremental delivery response format. It is based on [the proposed changes to `@graphql-tools/utils`](https://github.com/ardatan/graphql-tools/pull/6243), that library being the inspiration for the most recent changes within this repository in #3580.

I am not sure what the testing story will be for this PR as I am not too familiar with development with this repository. Feel free to point me in the right direction or to take this PR over. :)

### CODE REVIEW

**Comment by yaacovCR** [comment]
I have been unable to end to end tests and added v17 to the testing matrix. Unfortunately, there are some build errors that seem to be possibly on Main or possibly some sort of local development error on my side.

It also appears that there are some new test errors with the language server possibly with loaders from graphql utils not playing nicely with v17, but tricky for me to debug…

All end to end tests are passing however, so this is a start, but any help appreciated. :)

File: packages/graphiql/cypress/e2e/incremental-delivery.cy.ts
Comment (dimaMachina) [inline packages/graphiql/cypress/e2e/incremental-delivery.cy.ts:167]: only this test now fails

### Expected

image

### Received

image

File: packages/graphiql/test/e2e-server.js
Comment (dimaMachina) [inline packages/graphiql/test/e2e-server.js:23]: I think the reason why you use `graphql-helix` because it exports helper functions `getGraphQLParameters`, `processRequest` and  `sendResult `?

Is there another library which we can use instead? because it's kinda of legacy and the last update of `graphql-helix` was 2 years ago

File: packages/graphiql/test/schema.js
Comment (dimaMachina) [inline packages/graphiql/test/schema.js:31]:
```suggestion
const enableExperimentalIncrementalDelivery = parseInt(version) > 16
```
maybe?
image

File: packages/graphql-language-service-server/src/__tests__/MessageProcessor.spec.ts
Comment (dimaMachina) [inline packages/graphql-language-service-server/src/__tests__/MessageProcessor.spec.ts:366]:
```suggestion
    const enableExperimentalIncrementalDelivery = parseInt(version) > 16;
```

File: packages/graphql-language-service/src/interface/__tests__/getDiagnostics-test.ts
Comment (dimaMachina) [inline packages/graphql-language-service/src/interface/__tests__/getDiagnostics-test.ts:96]:
```suggestion
                  parseInt(version) > 16
                    ? { nodes: definition }
                    : definition,
```

File: packages/graphql-language-service/src/interface/__tests__/getDiagnostics-test.ts
Comment (dimaMachina) [inline packages/graphql-language-service/src/interface/__tests__/getDiagnostics-test.ts:174]:
```suggestion
              parseInt(version) > 16
                ? { nodes: node.name }
                : node.name,
```

File: packages/graphql-language-service/src/utils/__tests__/validateWithCustomRules-test.ts
Comment (dimaMachina) [inline packages/graphql-language-service/src/utils/__tests__/validateWithCustomRules-test.ts:44]:
```suggestion
                parseInt(version) > 16
                  ? { nodes: node }
                  : node,
```

File: packages/monaco-graphql/test/monaco-editor.test.ts
Comment (dimaMachina) [inline packages/monaco-graphql/test/monaco-editor.test.ts:18]:
```suggestion
      `✓ ${parseInt(version) > 16 ? 856 : 843} modules transformed.`,
```

File: packages/vscode-graphql/package.json
Comment (dimaMachina) [inline packages/vscode-graphql/package.json:182]:
```suggestion
    "graphql": "^16.8.1 || ^17.0.0-alpha.5",
```
to match other updates in monorepo

File: packages/graphql-language-service/package.json
Comment (dimaMachina) [inline packages/graphql-language-service/package.json:48]:
```suggestion
    "graphql": "^16.8.1 || ^17.0.0-alpha.5",
```

Comment (dimaMachina) [reply packages/graphql-language-service/package.json:48]: it's `devDependency` and not peer

File: packages/vscode-graphql-execution/package.json
Comment (dimaMachina) [reply packages/vscode-graphql-execution/package.json:112]: i was wrong

File: packages/graphiql/test/e2e-server.js
Comment (yaacovCR) [reply packages/graphiql/test/e2e-server.js:23]: existing `graphql-http` does not support multipart requests and gives me errors
then i tried `graphql-yoga`, but when I added it, jest mocks stopped working, specifically:

https://github.com/graphql/graphiql/blob/1c1d2b582809cb7ad86d8bf197af63cd3d5025b7/packages/graphql-language-service-server/src/__tests__/GraphQLCache.test.ts#L12

https://github.com/graphql/graphiql/blob/1c1d2b582809cb7ad86d8bf197af63cd3d5025b7/packages/graphql-language-service-server/src/__tests__/MessageProcessor.spec.ts#L20

And I was not able to fix that.

I suppose it is related to the fact that `@whatwg-node/fetch'` is a dependency of `graphql-yoga` , and perhaps the fact that `@whatwg-node/fetch'` is loaded by yoga via esm? Or maybe just first? I actually have no idea.

`graphql-helix` has no dependencies and so that's good. I look at it as temporary until `graphql-http` is upgraded to support multipart requests for incremental delivery.

File: packages/graphiql/cypress/e2e/incremental-delivery.cy.ts
Comment (dimaMachina) [inline packages/graphiql/cypress/e2e/incremental-delivery.cy.ts:5]:
```suggestion
if (parseInt(version) < 17) {
```

File: packages/codemirror-graphql/src/__tests__/lint-test.ts
Comment (dimaMachina) [inline packages/codemirror-graphql/src/__tests__/lint-test.ts:70]:
```suggestion
              parseInt(version) > 16
                ? { nodes: node }
                : node,
```

File: packages/graphiql/cypress/e2e/incremental-delivery.cy.ts
Comment (yaacovCR) [reply packages/graphiql/cypress/e2e/incremental-delivery.cy.ts:167]: ff8fdddfad9289b08ab61df4f88378e207e9e731

Comment (yaacovCR) [reply packages/graphiql/cypress/e2e/incremental-delivery.cy.ts:167]: hmm the patch-package works for me locally, but not for the unit tests... hopefully we can get another alpha released. canary builds on graphql-js appear broken

Comment (dimaMachina) [reply packages/graphiql/cypress/e2e/incremental-delivery.cy.ts:167]: seems didn't help, still have same result 
image

File: packages/vscode-graphql-execution/package.json
Comment (yaacovCR) [reply packages/vscode-graphql-execution/package.json:112]: I guess my idea was that for the devDependency we use the latest alpha, but in theory we support from alpha.2 onward, so when it's a dependency, we can use the range -- but skipping the very first alpha because it doesn't support incremental delivery anyway and complicated the version checking.

Not sure that's necessary, defer to you.

File: packages/vscode-graphql/package.json
Comment (dimaMachina) [inline packages/vscode-graphql/package.json:182]:
```suggestion
    "graphql": "^16.8.1 || ^17.0.0-alpha.5",
```
better to points on the working version?

File: packages/vscode-graphql-execution/package.json
Comment (dimaMachina) [reply packages/vscode-graphql-execution/package.json:112]: ok, let's keep `^17.0.0-alpha.2`

File: packages/graphiql/test/schema.js
Comment (yaacovCR) [reply packages/graphiql/test/schema.js]: 70bbc6cd6559c051f129a3b299806e01ede96c08

File: packages/graphql-language-service/src/interface/__tests__/getDiagnostics-test.ts
Comment (yaacovCR) [reply packages/graphql-language-service/src/interface/__tests__/getDiagnostics-test.ts]: 70bbc6cd6559c051f129a3b299806e01ede96c08

File: packages/graphql-language-service/src/interface/__tests__/getDiagnostics-test.ts
Comment (yaacovCR) [reply packages/graphql-language-service/src/interface/__tests__/getDiagnostics-test.ts]: 70bbc6cd6559c051f129a3b299806e01ede96c08

File: packages/monaco-graphql/test/monaco-editor.test.ts
Comment (yaacovCR) [reply packages/monaco-graphql/test/monaco-editor.test.ts]: 70bbc6cd6559c051f129a3b299806e01ede96c08

File: packages/graphiql/cypress/e2e/incremental-delivery.cy.ts
Comment (yaacovCR) [reply packages/graphiql/cypress/e2e/incremental-delivery.cy.ts]: 70bbc6cd6559c051f129a3b299806e01ede96c08

File: packages/codemirror-graphql/src/__tests__/lint-test.ts
Comment (yaacovCR) [reply packages/codemirror-graphql/src/__tests__/lint-test.ts]: 70bbc6cd6559c051f129a3b299806e01ede96c08

File: packages/graphiql/test/e2e-server.js
Comment (yaacovCR) [reply packages/graphiql/test/e2e-server.js]: 70bbc6cd6559c051f129a3b299806e01ede96c08

File: packages/vscode-graphql/package.json
Comment (yaacovCR) [reply packages/vscode-graphql/package.json:182]: yeah, alpha.2 should work as well, it just has the old duplicating format, which was introduced in alpha.3

alpha.5 (and probably alpha.4) currently has a bug with defer nested under stream

File: packages/graphql-language-service/src/utils/__tests__/validateWithCustomRules-test.ts
Comment (yaacovCR) [reply packages/graphql-language-service/src/utils/__tests__/validateWithCustomRules-test.ts]: 70bbc6cd6559c051f129a3b299806e01ede96c08

File: packages/graphql-language-service-server/src/__tests__/MessageProcessor.spec.ts
Comment (yaacovCR) [reply packages/graphql-language-service-server/src/__tests__/MessageProcessor.spec.ts]: 70bbc6cd6559c051f129a3b299806e01ede96c08

File: packages/graphiql/cypress/e2e/incremental-delivery.cy.ts
Comment (yaacovCR) [reply packages/graphiql/cypress/e2e/incremental-delivery.cy.ts:167]: had the patch in the wrong folder, now it's working in our tests, although any servers using alpha.4 or alpha.5 are going to have the bug unless they also patch :(

**Comment by yaacovCR** [comment]
> ### 🦋 Changeset detected
> Latest commit: [205980c](https://github.com/graphql/graphiql/commit/205980c1b860b06917dd9eac17e10db59bbf3e3e)
> 
> **The changes in this PR will be included in the next version bump.**
> 
> This PR includes changesets to release 4 packages
> Name	Type
> @graphiql/react	Minor
> graphiql	Minor
> @graphiql/plugin-code-exporter	Major
> @graphiql/plugin-explorer	Major
> Not sure what this means? [Click here to learn what changesets are](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md).
> 
> [Click here if you're a maintainer who wants to add another changeset to this PR](https://github.com/yaacovCR/graphiql/new/new-incremental-format?filename=.changeset/purple-chairs-know.md&value=---%0A%22cm6-graphql%22%3A%20patch%0A%22codemirror-graphql%22%3A%20patch%0A%22%40graphiql%2Fplugin-code-exporter%22%3A%20patch%0A%22%40graphiql%2Fplugin-explorer%22%3A%20patch%0A%22%40graphiql%2Freact%22%3A%20patch%0A%22%40graphiql%2Ftoolkit%22%3A%20patch%0A%22graphiql%22%3A%20patch%0A%22graphql-language-service-cli%22%3A%20patch%0A%22graphql-language-service-server%22%3A%20patch%0A%22graphql-language-service%22%3A%20patch%0A%22monaco-graphql%22%3A%20patch%0A%22vscode-graphql-execution%22%3A%20patch%0A---%0A%0Asupport%20new%20incremental%20delivery%20format%0A)

not sure what these other packages are and why they are going to be released as a breaking change?

File: packages/graphiql-react/src/execution.tsx
Comment (yaacovCR) [reply packages/graphiql-react/src/execution.tsx]: removed!

**Review by dimaMachina — APPROVED** [review]

**Comment by acao** [comment]
@dimaMachina next time, please do not merge a PR like this without asking. I need to manually revert many changes for us to return to spec compliance

File: resources/patches/graphql+17.0.0-alpha.5.patch
Comment (acao) [inline resources/patches/graphql+17.0.0-alpha.5.patch:1]: @dimaMachina again, a reminder that this patch is being published to users. we should not be patching `graphql` before publishing as a reference ecosystem

Comment (dimaMachina) [reply resources/patches/graphql+17.0.0-alpha.5.patch:1]: This patch was not published to users

**Pull request merged**
